sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs

```diff
@@ -0,0 +1,130 @@
+    /// </summary>
+    internal partial class WebPubSubAuthenticationPolicy : HttpPipelineSynchronousPolicy
+    {
+        private const int MaxTokenLength = 4096;
+
+        /// <summary>
+        /// Generates client bearwer token.
+        /// </summary>
+        /// <param name="audience"></param>
+        /// <param name="claims"></param>
+        /// <param name="key"></param>
+        /// <param name="expireAfter"></param>
+        /// <param name="hmacSha512">SHA512 if true, otherwise SHA256</param>
+        /// <returns></returns>
+        public static string GenerateAccessToken(
+            string audience,
+            IEnumerable<Claim> claims,
+            AzureKeyCredential key,
+            TimeSpan expireAfter = default,
+            bool hmacSha512 = false)
+        {
+            if (expireAfter == default) expireAfter = TimeSpan.FromMinutes(10);
+
+            var jwtToken = JwtUtils.GenerateJwtBearer(
+                audience: audience,
+                claims: claims,
+                expiresAfter: expireAfter,
+                key: key,
+                hmacSha512: hmacSha512
+            );
+
+            Debug.Assert(jwtToken.Length <= MaxTokenLength);
+
+            return jwtToken;
+        }
+
+        private static class JwtUtils
+        {
+            private static readonly JwtSecurityTokenHandler JwtTokenHandler = new JwtSecurityTokenHandler();
+            private static readonly ConditionalWeakTable<AzureKeyCredential, UniqueKey> KeyForCredential = new ConditionalWeakTable<AzureKeyCredential, UniqueKey>();
+
+            //public static string GenerateAccessToken(byte[] key, string audience, string userId, TimeSpan expiresAfter)
+            //{
+            //    var expiresAt = DateTime.UtcNow.Add(expiresAfter);
+
+            //    Claim[] claims = null;
+            //    if (userId != null)
+            //    {
+            //        claims = new Claim[] { new Claim(ClaimTypes.NameIdentifier, userId) };
+            //    }
+
+            //    var securityKey = new SymmetricSecurityKey(key);
+            //    var credentials = new SigningCredentials(securityKey, SecurityAlgorithms.HmacSha256);
+
+            //    var token = JwtTokenHandler.CreateJwtSecurityToken(
+            //        issuer: null,
+            //        audience: audience,
+            //        subject: claims == null ? null : new ClaimsIdentity(claims),
+            //        expires: expiresAt,
+            //        signingCredentials: credentials);
+
+            //    return JwtTokenHandler.WriteToken(token);
+            //}
+
+            public static string GenerateJwtBearer(
+                string audience,
+                IEnumerable<Claim> claims,
+                TimeSpan expiresAfter,
+                AzureKeyCredential key,
+                string issuer = null,
+                bool hmacSha512 = false)
+            {
+                var expiresAt = DateTime.UtcNow.Add(expiresAfter);
```

Comment:
Should we allow folks to specify `ExpiresOn` as well so they don't have to worry about clock skew?

Reply:
Well, this is used for the 10 minutes you commented on above. This is not something exposed to the user.

Reply:
Isn't it also used for `GetClientAccessUri`?  That's the case I'm more concerned about.

Reply:
True,

@schaabs, what do you recommend here? I would prefer not to have two overloads: one taking expiresAfter and one taking expiresOn

Reply:
This could just be a single overload with `expiresOn` and you do the math upstream by saying `DateTime.UtcNow.AddMinutes(10)`.  The only place we'd really want/need two overloads is `GetClientAccessUri` if we think `I want an access URI that ends at exactly 12 midnight PST for compliance reasons` is an important use case.

Reply:
Yeah, but then the user has to do the math. Is that what we want? Is that what we do in all libraries? I feel like it would be good to decide at SDK level. I will file an issue to resolve it after beta.

Reply:
To be clearer, I'm suggesting a single code path through all your internals with an absolute time and then an overload on `DateTimeOffset`/`TimeSpan` at the public `GetClientAccessUri` boundary to do the math for the customer.  We can consider that for post-beta though.